### DISCUSSION TITLE
STIG Manager with NSS token instead of CAC?

### DISCUSSION THREAD

**Discussion by st3v1s** [NUWCDIVNPT/stig-manager#1046] [Q&A]
If we want to use STIG Manager on another network and employ the use of NSS pki cert, would I just need to replace across the board where I have used the Certificates_PKCS7_v5.9_DoD.pem certificate for CAC cards? And put in its place the chain certificate for NSS token certificates?

I am using the stigman-orchestration for CAC as my template.

Thanks for any assistance you can provide.

Steve

**Comment by cd-rite** [comment]
In theory, I think it could be that simple!  I suppose there is the possibility that nss tokens have different fields you could be mapping somewhere in keycloak, but they should also have the standard ones in common with CAC....

**Reply by st3v1s** [reply]
So we did just that and replaced instances of of the DoD CAC cert chain with the NSS cert chain.  Created a new truststore.p12 using the new cert chain file.

Ran the docker compose and all containers started.  Was able to access the Keycloak admin page with no issue.  Upon attempting to go to https://fqdn/stigman/, we are asked for cert, then see error "Ext.Ajax.request() failed".

Any ideas what I may be missing?

**Reply by cd-rite** [reply]
Hmmm does your browser trust the cert being used? The machine your are requesting from will need to know about the new certs you are using….
Which version of STIGMan are you using? The latest (1.3.8) includes a more helpful warning when not operating in a secure context (ie. The browser does not trust the cert of the machine your are connecting to, or when using a self-signed cert ) .

**Reply by st3v1s** [reply]
So to elaborate the error pops up immediateley after enter the token PIN and selecting the user profile name....

As for the browser cert trust, should we import the new cert chain that we are using?

**Comment by st3v1s** [comment]
The stigman container log shows "unable to verify first certificate"

**Comment by csmig** [comment]
If you require trust of CAs that are not included in the base Node.js, try specifying this envvar when running the STIGMAN container:

`NODE_EXTRA_CA_CERTS=[file-with-your-nss-certificate-chain]`

https://nodejs.org/api/cli.html#node_extra_ca_certsfile

PS. the fragment URL above doesn't always scroll to the proper page location, you may have to search within the page manually)

**Comment by csmig** [comment]
To further clarify, the envvar `NODE_EXTRA_CA_CERTS` is only needed if you configure an HTTPS URL for `STIGMAN_OIDC_PROVIDER`.

Many containerized deployments configure the STIGMAN API to connect to Keycloak using a back channel, non-TLS connection. In that case there is no need for the extra CA certs in the API setup. YMMV.

**Reply by st3v1s** [reply]
ah, ok...in our setup the STIGMAN_OIDC_PROVIDER is http, but the STIGMAN_CLIENT_OIDC_PROVIDER is https.

We tried it but got same Ajax error

**Reply by csmig** [reply]
If the API is generating that message about failing to verify the first cert, then it is trying to make an outbound TLS connection and failing. The only connection the API makes is to Keycloak during API bootstrap. So something isn't configured as it seems.

Perhaps sharing your full orchestration config would be helpful. Maybe the container log, as well.

**Reply by st3v1s** [reply]
# STIG Manager with CAC Authentication

version: '2.1'

services:
  nginx:
    #image: nginx:1.23.1
    # alternative image from Ironbank
    image: registry1.dso.mil/ironbank/opensource/nginx/nginx:1.23.1
    volumes:
      - /etc/reverse/nginx/nginx.conf:/etc/nginx/nginx.conf
      - /etc/reverse/nginx/fqdn_cer.pem:/etc/nginx/cert.pem
      - /etc/reverse/nginx/fqdn_key.pem:/etc/nginx/privkey.pem
      - /etc/reverse/nginx/nss-certs.pem:/etc/nginx/nss-certs.pem
      - /etc/reverse/nginx/index.html:/usr/share/nginx/html/index.html
    ports:
      - "443:443"
  keycloak:
    # alternative image from Ironbank
    image: registry1.dso.mil/ironbank/opensource/keycloak/keycloak:19.0.2
    # image: quay.io/keycloak/keycloak:19.0.2
    ports:
      - 8080:8080
      - 8443:8443
    environment:
      - PROXY_ADDRESS_FORWARDING=true
      - KEYCLOAK_ADMIN=admin
      - KEYCLOAK_ADMIN_PASSWORD=password
      - KC_PROXY=edge
      - KC_HOSTNAME_URL=https://fqdn/kc/
      - KC_HOSTNAME_ADMIN_URL=https://fqdn/kc/
      - KC_SPI_X509CERT_LOOKUP_PROVIDER=nginx
      - KC_SPI_X509CERT_LOOKUP_NGINX_SSL_CLIENT_CERT=SSL-CLIENT-CERT
      - KC_SPI_TRUSTSTORE_FILE_FILE=/tmp/truststore.p12
      - KC_SPI_TRUSTSTORE_FILE_PASSWORD=trust
      - KC_HTTPS_CERTIFICATE_FILE=/opt/keycloak/nss-certs.pem
      - KC_METRICS_ENABLED=true
      - KC_HEALTH_ENABLED=true
    command: start --import-realm
    volumes:
      - /etc/reverse/kc/truststore.p12:/tmp/truststore.p12
      - /etc/reverse/kc/stigman_realm.json:/opt/keycloak/data/import/stigman_realm.json
      - /etc/reverse/kc/create-x509-user.jar:/opt/keycloak/providers/create-x509-user.jar
      - /etc/reverse/kc/nss-certs.pem:/opt/keycloak/nss-certs.pem
      # uncomment below to persist Keycloak data
      # - ./kc/h2:/opt/keycloak/data/h2
  stigman:
    # alternative image based on Ironbank Node.js
    image: nuwcdivnpt/stig-manager:latest-ironbank
    # image: nuwcdivnpt/stig-manager:latest
    ports:
      - 54000:54000
    environment:
      - STIGMAN_OIDC_PROVIDER=http://keycloak:8080/realms/stigman
      - STIGMAN_CLIENT_OIDC_PROVIDER=https://fqdn/kc/realms/stigman
      - STIGMAN_CLASSIFICATION=U
      - STIGMAN_DB_HOST=fqdn
      - STIGMAN_DB_PORT=50001
      - STIGMAN_API_ADDRESS=fqdn
      - STIGMAN_DB_PASSWORD=stigman
      # uncomment below to fetch current STIG library from DISA and import it
      # - STIGMAN_INIT_IMPORT_STIGS=true
    init: true
  mysql:
    # image: mysql:8.0.21
    # alternative image from Ironbank
    image: registry1.dso.mil/ironbank/opensource/mysql/mysql8:8.0.21
    ports: 
      - 50001:3306
    command: --innodb-buffer-pool-size=512M --sort_buffer_size=64M 
    environment:
      - MYSQL_ROOT_PASSWORD=rootpw
      - MYSQL_USER=stigman
      - MYSQL_DATABASE=stigman
      - MYSQL_PASSWORD=stigman
    # uncomment below to persist MySQL data
    # volumes:
    #  - ./mysql-data:/var/lib/mysql

**Reply by st3v1s** [reply]
I can't share container log as its on another network, but this is very similar to what our docker-compose.yml file looks like.

**Comment by csmig** [comment]
Thanks. The container log is required to troubleshoot the error on the API. At a minimum, we need to see the context in which that `failed to verify first certificate` error occurs. I have a suspicion on what is going on, but need the log to go further.

If the verify error is happening after the API successfully starts (it has emitted the log record with `{"type": "started"}`) then I suspect the `jwksUri` value returned by Keycloak when it is probed for its openid-configuration is not proper.  Please check for  the very early log record that looks like:

`{"date":"2023-08-04T20:58:45.132Z","level":3,"component":"oidc","type":"discovery","data":{"success":true,"metadataUri":"http://keycloak:8080/realms/stigman/.well-known/openid-configuration","jwksUri":"http://keycloak:8080/realms/stigman/protocol/openid-connect/certs"}}`

The `metadataUri` and  `jwksUri` values should  have the same scheme (either HTTP or HTTPS), hostname, and path prefix. If they do not, then Keycloak is at fault. Given your compose file, they should both start with `http://keycloak:8080/realms/stigman`. I'm wondering if the  `jwksUri` value is `https://fqdn/kc/realms/stigman/protocol/openid-connect/certs`, which requires the NODE_EXTRA_CA_CERTS be set.  I don't see the NODE_EXTRA_CA_CERTS value in your compose file, but it would be needed in this anomalous case.

Ideally we would make our first fetch from the jwksUri during bootstrap, so a failure could be reported early and the API would not start. But we currently defer the first call to the jwksUri until the first endpoint request is made, usually when you fire up  the web app, which is consistent with your experience.

**Reply by st3v1s** [reply]
So I checked the logs this morning and saw "socket hang up" and as you suspected metadatauri and jwksuri are http and https respectivley.  I commented out the STIGMAN_CLIENT_OIDC_PROVIDER_URL and the NODE_EXTRA_CA_CERTS and although I am asked for my NSS pin to get to keycloak I still login with admin password.  As for stigman, I see new error "failed to get http://fqdn:8080/realms/stigman/.well-known/openid-configuration     Any ideas?

**Reply by st3v1s** [reply]
Removed the comments mentioned above and back to ajax error....in the logs I see : docker-stiman-1 (error: error in secret or public key callback: unable to verify the first certificate, stack, Uauthorized: error in secret or public key callback: unable to verify the first certificate at HttpError.create  (/home/node/node_modules/express-openapi-validator/dist/framework/types.js

then docker-nginx-1 client closed connection while waiting for request

**Comment by st3v1s** [comment]
BOOM!   got it working!   so I had to create a volume to map the NODE_EXTRA_CA_CERTS value to a location inside the container so it could find it.  It dawned on me that I had set it to the local machine and not within the container.

**Comment by csmig** [comment]
Great news you've got things working! However, your Keycloak is not behaving as expected and it would be helpful to understand why. I'll run some tests in my lab using different versions. If you are able, please share all the envvars you've applied to Keycloak and the version you're running. Thanks!

When Keycloak OIDC metadata is fetched using the backend URL as you've apparently configured the API to do, most (all?) the endpoints should reflect the backend scheme and host. For example, when the API requests metadata from `http://keycloak:8080/realms/stigman/.well-known/openid-configuration`, we expect the value for the jwksUri should be `http://keycloak:8080/realms/stigman/protocol/openid-connect/certs`. But you're seeing the frontend scheme and host in that value, which is unexpected and perhaps a bug in Keycloak.

**Comment by st3v1s** [comment]
I am using IronBank Keycloak version 19.0.2, here are Keycloak variables I'm using:
keycloak:
# alternative image from Ironbank
image: registry1.dso.mil/ironbank/opensource/keycloak/keycloak:19.0.2
# image: quay.io/keycloak/keycloak:19.0.2
ports:
- 8080:8080
- 8443:8443
environment:
- PROXY_ADDRESS_FORWARDING=true
- KEYCLOAK_ADMIN=admin
- KEYCLOAK_ADMIN_PASSWORD=password
- KC_PROXY=edge
- KC_HOSTNAME_URL=https://fqdn/kc/
- KC_HOSTNAME_ADMIN_URL=https://fqdn/kc/
- KC_SPI_X509CERT_LOOKUP_PROVIDER=nginx
- KC_SPI_X509CERT_LOOKUP_NGINX_SSL_CLIENT_CERT=SSL-CLIENT-CERT
- KC_SPI_TRUSTSTORE_FILE_FILE=/tmp/truststore.p12
- KC_SPI_TRUSTSTORE_FILE_PASSWORD=trust
- KC_HTTPS_CERTIFICATE_FILE=/opt/keycloak/nss-certs.pem
- KC_METRICS_ENABLED=true
- KC_HEALTH_ENABLED=true
command: start --import-realm
volumes:
- /etc/reverse/kc/truststore.p12:/tmp/truststore.p12
- /etc/reverse/kc/stigman_realm.json:/opt/keycloak/data/import/stigman_realm.json
- /etc/reverse/kc/create-x509-user.jar:/opt/keycloak/providers/create-x509-user.jar
- /etc/reverse/kc/nss-certs.pem:/opt/keycloak/nss-certs.pem

**Reply by st3v1s** [reply]
New Question:  If I just wanted to remove the CAC card authentication piece, but keep the reverse proxy so that http redirects to https, what would I need to remove?  Authentication would be just username/password.

Thanks for any insight you can provide!  I need this setup for another network that does not have smartcard but must have https.

**Reply by cd-rite** [reply]
Hi @st3v1s I think @jeremyatourville may have a solution for you....
In another discussion, he mentioned he had created a setup like what you are looking for, based on a fork of our stigman-orchestration repo.

His repo is here: 
https://github.com/jeremyatourville/stigman-orchestration

I've added a new section to our Watcher docs to list "Community Solutions" that other users have made available:
https://github.com/NUWCDIVNPT/stigman-orchestration#community-solutions

**Reply by st3v1s** [reply]
Thanks, I got it working, but had to remove the x509 browser flow in stigman realm and replace with browser (built-in) browser based authentication.  Otherwise it would prompt me for the CAC card certificate.

**Reply by jeremyatourville** [reply]
@st3v1s - Interesting, I don't recall needing to make any changes in the stigman realm.  I just added user accounts with username and password.  What does your final docker-compose file look like?

**Reply by st3v1s** [reply]
@jeremyatourville the other interesting thing is that I get prompted for CAC card upon going to the keycloak page, but if I hit Cancel its not a problem.

Below is my docker-compose file:

version: '2.1'

services:
nginx:
image: nginx:1.23.1
volumes:
- /etc/reverse/nginx/nginx.conf:/etc/nginx/nginx.conf
- /etc/reverse/nginx/fqdn_cer.pem:/etc/nginx/cert.pem
- /etc/reverse/nginx/fqdn_key.pem:/etc/nginx/privkey.pem
- /etc/reverse/nginx/dod-certs.pem:/etc/nginx/dod-certs.pem
- /etc/reverse/nginx/index.html:/usr/share/nginx/html/index.html
ports:
- "443:443"
keycloak:
image: registry1.dso.mil/ironbank/opensource/keycloak/keycloak:19.0.2
ports:
- 8080:8080
- 8443:8443
environment:
- PROXY_ADDRESS_FORWARDING=true
- KEYCLOAK_ADMIN=admin
- KEYCLOAK_ADMIN_PASSWORD=password
- KC_PROXY=edge
- KC_HOSTNAME=fqdn
- KC_HOSTNAME_ADMIN_URL=https://fqdn/kc
- KC_SPI_TRUSTSTORE_FILE_FILE=/tmp/truststore.p12
- KC_SPI_TRUSTSTORE_FILE_PASSWORD=password
command: start --import-realm
volumes:
- /etc/reverse/kc/truststore.p12:/tmp/truststore.p12
- /etc/reverse/kc/stigman_realm.json:/opt/keycloak/data/import/stigman_realm.json
# uncomment below to persist Keycloak data
# - ./kc/h2:/opt/keycloak/data/h2
stigman:
image: nuwcdivnpt/stig-manager:latest
ports:
- 54000:54000
environment:
- STIGMAN_OIDC_PROVIDER=http://keycloak:8080/realms/stigman
- STIGMAN_CLIENT_OIDC_PROVIDER=https://fqdn/kc/realms/stigman
- STIGMAN_CLASSIFICATION=U
- STIGMAN_DB_HOST=fqdn
- STIGMAN_DB_PORT=50001
- STIGMAN_API_ADDRESS=fqdn
- STIGMAN_DB_USER=stigman
- STIGMAN_DB_PASSWORD=stigman
# uncomment below to fetch current STIG library from DISA and import it
# - STIGMAN_INIT_IMPORT_STIGS=true
init: true
mysql:
image: mysql:8.0.21
ports:
- 50001:3306
command: --innodb-buffer-pool-size=512M --sort_buffer_size=64M
environment:
- MYSQL_ROOT_PASSWORD=rootpw
- MYSQL_USER=stigman
- MYSQL_DATABASE=stigman
- MYSQL_PASSWORD=stigman
# uncomment below to persist MySQL data
# volumes:
# - ./mysql-data:/var/lib/mysql

**Reply by jeremyatourville** [reply]
It generally looks ok.  A little hard to read easily though.  Please use the add code button `<>`    
Anyhow, one thing I can think of is running a docker prune.  I have seen containers hold on to old settings sometimes.  As long as you have everything backed by volumes you shouldn't lose any of your data.  

I don't see this volume: ` - ./kc/create-x509-user.jar:/opt/keycloak/providers/create-x509-user.jar` which is a GOOD thing if you don't want to use CAC auth.

**Reply by st3v1s** [reply]
Did you change the original stigman_realm.json file?

**Reply by jeremyatourville** [reply]
no
